core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocateUnassignedDecision.java

```diff
@@ -268,6 +269,18 @@ public String getExplanation() {
         }
     }
 
+    private boolean hasNodeWithStaleOrCorruptShard() {
+        if (getNodeDecisions() != null && getNodeDecisions().isEmpty() == false) {
+            return getNodeDecisions().stream()
+                       .filter(result -> result.getShardStoreInfo() != null
```

Comment:
there is a method called `anyMatch(...)`. This will also make it return true on the first occurrence found